File: backend/src/websockets/protocol/requests.py
from pydantic import BaseModel

from backend.src.engine.game import PlayerId
from backend.src.models.ship_status import ShipStatus
from backend.src.websockets.protocol.message_types import Request, RequestTypes


class ShipToPlace(BaseModel):
    ship: ShipStatus
    row: int
    col: int
    horizontal: bool


class CreateGameRequest(Request):
    type: RequestTypes = RequestTypes.CREATE
    player_id: PlayerId


class JoinGameRequest(Request):
    type: RequestTypes = RequestTypes.JOIN
    player_id: PlayerId
    code: str


class GetStateRequest(Request):
    type: RequestTypes = RequestTypes.GET_STATE


class PlaceRandomRequest(Request):
    type: RequestTypes = RequestTypes.PLACE_RANDOM
    override: bool


class FireRequest(Request):
    type: RequestTypes = RequestTypes.FIRE
    row: int
    col: int


class ChatRequest(Request):
    type: RequestTypes = RequestTypes.CHAT
    message: str


class PlaceShipsRequest(Request):
    type: RequestTypes = RequestTypes.PLACE
    ships: list[ShipToPlace]

File: frontend/src/protocol/Requests.ts
import type {Request} from "./MessageType.ts";
import type {ShipToPlace} from "../types/ShipToPlace.ts";

export interface CreateGameRequest extends Request {
    player_id: string;
}

export interface JoinGameRequest extends Request {
    player_id: string;
    code: string;
}

export interface GetStateRequest extends Request {}

export interface PlaceRandomRequest extends Request {
    override: boolean; // TODO toujours true pour le moment, anyway le backend le supporte pas
}

export interface FireRequest extends Request {
    row: number;
    col: number;
}

export interface ChatRequest extends Request {
    message: string;
}

export interface PlaceShipsRequest extends Request {
    ships: ShipToPlace[];
}